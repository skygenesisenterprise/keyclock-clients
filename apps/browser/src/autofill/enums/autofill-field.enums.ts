export const AutofillFieldQualifier = {
  password: "password",
  newPassword: "newPassword",
  username: "username",
  cardholderName: "cardholderName",
  cardNumber: "cardNumber",
  cardExpirationMonth: "cardExpirationMonth",
  cardExpirationYear: "cardExpirationYear",
  cardExpirationDate: "cardExpirationDate",
  cardCvv: "cardCvv",
  identityTitle: "identityTitle",
  identityFirstName: "identityFirstName",
  identityMiddleName: "identityMiddleName",
  identityLastName: "identityLastName",
  identityFullName: "identityFullName",
  identityAddress1: "identityAddress1",
  identityAddress2: "identityAddress2",
  identityAddress3: "identityAddress3",
  identityCity: "identityCity",
  identityState: "identityState",
  identityPostalCode: "identityPostalCode",
  identityCountry: "identityCountry",
  identityCompany: "identityCompany",
  identityPhone: "identityPhone",
  identityEmail: "identityEmail",
  identityUsername: "identityUsername",
} as const;

export type AutofillFieldQualifierType =
  (typeof AutofillFieldQualifier)[keyof typeof AutofillFieldQualifier];
