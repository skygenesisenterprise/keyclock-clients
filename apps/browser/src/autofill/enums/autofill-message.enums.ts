export const AutofillMessageCommand = {
  collectPageDetails: "collectPageDetails",
  collectPageDetailsResponse: "collectPageDetailsResponse",
} as const;

export type AutofillMessageCommandType =
  (typeof AutofillMessageCommand)[keyof typeof AutofillMessageCommand];

export const AutofillMessageSender = {
  collectPageDetailsFromTabObservable: "collectPageDetailsFromTabObservable",
} as const;

export type AutofillMessageSenderType =
  (typeof AutofillMessageSender)[keyof typeof AutofillMessageSender];
