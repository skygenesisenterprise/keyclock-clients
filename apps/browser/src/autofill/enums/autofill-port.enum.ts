const AutofillPort = {
  InjectedScript: "autofill-injected-script-port",
} as const;

export { AutofillPort };
