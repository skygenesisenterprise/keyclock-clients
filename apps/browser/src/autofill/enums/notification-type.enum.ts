const NotificationType = {
  AddLogin: "add",
  ChangePassword: "change",
  UnlockVault: "unlock",
  AtRiskPassword: "at-risk-password",
} as const;

type NotificationTypes = (typeof NotificationType)[keyof typeof NotificationType];

export { NotificationType, NotificationTypes };
