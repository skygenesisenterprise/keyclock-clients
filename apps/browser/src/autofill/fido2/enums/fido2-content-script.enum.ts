export const Fido2ContentScript = {
  PageScript: "content/fido2-page-script.js",
  PageScriptDelayAppend: "content/fido2-page-script-delay-append-mv2.js",
  ContentScript: "content/fido2-content-script.js",
} as const;

export const Fido2ContentScriptId = {
  PageScript: "fido2-page-script-registration",
  ContentScript: "fido2-content-script-registration",
} as const;
