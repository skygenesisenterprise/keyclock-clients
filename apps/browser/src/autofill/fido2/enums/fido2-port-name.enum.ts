export const Fido2PortName = {
  InjectedScript: "fido2-injected-content-script-port",
} as const;
