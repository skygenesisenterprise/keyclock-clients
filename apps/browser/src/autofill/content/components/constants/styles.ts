import { Theme } from "@bitwarden/common/platform/enums";

const lightTheme = {
  transparent: {
    hover: `rgb(0 0 0 / 0.02)`,
  },
  shadow: `rgba(168 179 200)`,
  primary: {
    100: `rgba(219, 229, 246)`,
    300: `rgba(121, 161, 233)`,
    600: `rgba(23, 93, 220)`,
    700: `rgba(26, 65, 172)`,
  },
  secondary: {
    100: `rgba(230, 233, 239)`,
    300: `rgba(168, 179, 200)`,
    500: `rgba(90, 109, 145)`,
    600: `rgba(83, 99, 131)`,
    700: `rgba(63, 75, 99)`,
  },
  success: {
    100: `rgba(219, 229, 246)`,
    600: `rgba(121, 161, 233)`,
    700: `rgba(26, 65, 172)`,
  },
  danger: {
    100: `rgba(255, 236, 239)`,
    600: `rgba(203, 38, 58)`,
    700: `rgba(149, 27, 42)`,
  },
  warning: {
    100: `rgba(255, 248, 228)`,
    600: `rgba(255, 191, 0)`,
    700: `rgba(172, 88, 0)`,
  },
  info: {
    100: `rgba(219, 229, 246)`,
    600: `rgba(121, 161, 233)`,
    700: `rgba(26, 65, 172)`,
  },
  art: {
    primary: `rgba(2, 15, 102)`,
    accent: `rgba(44, 221, 223)`,
  },
  text: {
    main: `rgba(27, 32, 41)`,
    muted: `rgba(90, 109, 145)`,
    contrast: `rgba(255, 255, 255)`,
    alt2: `rgba(255, 255, 255)`,
    code: `rgba(192, 17, 118)`,
  },
  background: {
    DEFAULT: `rgba(255, 255, 255)`,
    alt: `rgba(243, 246, 249)`,
    alt2: `rgba(23, 92, 219)`,
    alt3: `rgba(26, 65, 172)`,
    alt4: `rgba(2, 15, 102)`,
  },
  brandLogo: `rgba(23, 93, 220)`,
};

const darkTheme = {
  transparent: {
    hover: `rgb(255 255 255 / 0.02)`,
  },
  shadow: `rgba(0, 0, 0)`,
  primary: {
    100: `rgba(26, 39, 78)`,
    300: `rgba(26, 65, 172)`,
    600: `rgba(101, 171, 255)`,
    700: `rgba(170, 195, 239)`,
  },
  secondary: {
    100: `rgba(48, 57, 70)`,
    300: `rgba(82, 91, 106)`,
    500: `rgba(121, 128, 142)`,
    600: `rgba(143, 152, 166)`,
    700: `rgba(158, 167, 181)`,
  },
  success: {
    100: `rgba(11, 111, 21)`,
    600: `rgba(107, 241, 120)`,
    700: `rgba(191, 236, 195)`,
  },
  danger: {
    100: `rgba(149, 27, 42)`,
    600: `rgba(255, 78, 99)`,
    700: `rgba(255, 236, 239)`,
  },
  warning: {
    100: `rgba(172, 88, 0)`,
    600: `rgba(255, 191, 0)`,
    700: `rgba(255, 248, 228)`,
  },
  info: {
    100: `rgba(26, 65, 172)`,
    600: `rgba(121, 161, 233)`,
    700: `rgba(219, 229, 246)`,
  },
  art: {
    primary: `rgba(243, 246, 249)`,
    accent: `rgba(44, 221, 233)`,
  },
  text: {
    main: `rgba(243, 246, 249)`,
    muted: `rgba(136, 152, 181)`,
    contrast: `rgba(18 26 39)`,
    alt2: `rgba(255, 255, 255)`,
    code: `rgba(255, 143, 208)`,
  },
  background: {
    DEFAULT: `rgba(32, 39, 51)`,
    alt: `rgba(18, 26, 39)`,
    alt2: `rgba(47, 52, 61)`,
    alt3: `rgba(48, 57, 70)`,
    alt4: `rgba(18, 26, 39)`,
  },
  brandLogo: `rgba(255, 255, 255)`,
};

export const themes = {
  light: lightTheme,
  dark: darkTheme,

  // For compatibility
  system: lightTheme,
};

export const spacing = {
  4: `16px`,
  3: `12px`,
  2: `8px`,
  "1.5": `6px`,
  1: `4px`,
};

export const border = {
  radius: {
    large: `8px`,
    full: `9999px`,
  },
};

export const typography = {
  body1: `
    line-height: 24px;
    font-family: Roboto, sans-serif;
    font-size: 16px;
  `,
  body2: `
    line-height: 20px;
    font-family: Roboto, sans-serif;
    font-size: 14px;
  `,
  helperMedium: `
    line-height: 16px;
    font-family: Roboto, sans-serif;
    font-size: 12px;
  `,
};

export const ruleNames = {
  fill: "fill",
  stroke: "stroke",
} as const;

type RuleName = (typeof ruleNames)[keyof typeof ruleNames];

/*
 * `color` is an intentionally generic name here, since either fill or stroke may apply, due to
 * inconsistent SVG construction. This consequently precludes dynamic multi-colored icons here.
 */
export const buildIconColorRule = (color: string, rule: RuleName = ruleNames.fill) => `
  ${rule}: ${color};
`;

export const animations = {
  spin: `
    0% {
      transform: rotate(0deg);
    }
    100% {
      transform: rotate(359deg);
    }
  `,
};

export function scrollbarStyles(theme: Theme, color?: { thumb?: string; track?: string }) {
  const thumbColor = color?.thumb || themes[theme].secondary["500"];
  const trackColor = color?.track || themes[theme].background.alt;

  return {
    /* FireFox & Chrome support */
    default: `
      scrollbar-color: ${thumbColor} ${trackColor};
    `,
    /* Safari Support */
    safari: `
      ::-webkit-scrollbar {
        overflow: auto;
      }
      ::-webkit-scrollbar-thumb {
        border-width: 4px;
        border-style: solid;
        border-radius: 0.5rem;
        border-color: transparent;
        background-clip: content-box;
        background-color: ${thumbColor};
      }
      ::-webkit-scrollbar-track {
        ${trackColor};
      }
      ::-webkit-scrollbar-thumb:hover {
        ${themes[theme].secondary["600"]};
      }
    `,
  };
}
