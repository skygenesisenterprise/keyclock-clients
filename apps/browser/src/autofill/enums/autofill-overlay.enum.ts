import { CipherType } from "@bitwarden/common/vault/enums";

export const AutofillOverlayElement = {
  Button: "autofill-inline-menu-button",
  List: "autofill-inline-menu-list",
} as const;

export type AutofillOverlayElementType =
  (typeof AutofillOverlayElement)[keyof typeof AutofillOverlayElement];

export const AutofillOverlayPort = {
  Button: "autofill-inline-menu-button-port",
  ButtonMessageConnector: "autofill-inline-menu-button-message-connector",
  List: "autofill-inline-menu-list-port",
  ListMessageConnector: "autofill-inline-menu-list-message-connector",
} as const;

export const RedirectFocusDirection = {
  Current: "current",
  Previous: "previous",
  Next: "next",
} as const;

export const InlineMenuFillTypes = {
  AccountCreationUsername: 5,
  PasswordGeneration: 6,
  CurrentPasswordUpdate: 7,
} as const;

export type InlineMenuFillTypeValue =
  (typeof InlineMenuFillTypes)[keyof typeof InlineMenuFillTypes];

export type InlineMenuFillType = InlineMenuFillTypeValue | CipherType;

export const InlineMenuAccountCreationFieldType = {
  Text: "text",
  Email: "email",
  Password: "password",
  Totp: "totp",
} as const;

export type InlineMenuAccountCreationFieldTypes =
  (typeof InlineMenuAccountCreationFieldType)[keyof typeof InlineMenuAccountCreationFieldType];

export const MAX_SUB_FRAME_DEPTH = 8;
