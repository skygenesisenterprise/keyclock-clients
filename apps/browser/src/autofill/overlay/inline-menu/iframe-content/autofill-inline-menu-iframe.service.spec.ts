import { mock } from "jest-mock-extended";

import { EVENTS } from "@bitwarden/common/autofill/constants";
import { ThemeType } from "@bitwarden/common/platform/enums";

import { AutofillOverlayPort } from "../../../enums/autofill-overlay.enum";
import { createPortSpyMock } from "../../../spec/autofill-mocks";
import {
  flushPromises,
  sendPortMessage,
  triggerPortOnDisconnectEvent,
} from "../../../spec/testing-utils";

import { AutofillInlineMenuIframeService } from "./autofill-inline-menu-iframe.service";

describe("AutofillInlineMenuIframeService", () => {
  let autofillInlineMenuIframeService: AutofillInlineMenuIframeService;
  let portSpy: chrome.runtime.Port;
  let shadowAppendSpy: jest.SpyInstance;
  let handlePortDisconnectSpy: jest.SpyInstance;
  let handlePortMessageSpy: jest.SpyInstance;
  let sendExtensionMessageSpy: jest.SpyInstance;

  beforeEach(() => {
    const shadow = document.createElement("div").attachShadow({ mode: "open" });
    autofillInlineMenuIframeService = new AutofillInlineMenuIframeService(
      shadow,
      AutofillOverlayPort.Button,
      { height: "0px" },
      "title",
      "ariaAlert",
    );
    shadowAppendSpy = jest.spyOn(shadow, "appendChild");
    handlePortDisconnectSpy = jest.spyOn(
      autofillInlineMenuIframeService as any,
      "handlePortDisconnect",
    );
    handlePortMessageSpy = jest.spyOn(autofillInlineMenuIframeService as any, "handlePortMessage");
    chrome.runtime.connect = jest.fn((connectInfo: chrome.runtime.ConnectInfo) =>
      createPortSpyMock(connectInfo.name),
    ) as unknown as typeof chrome.runtime.connect;
    sendExtensionMessageSpy = jest.spyOn(
      autofillInlineMenuIframeService as any,
      "sendExtensionMessage",
    );
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe("initMenuIframe", () => {
    it("sets up the iframe's attributes", () => {
      autofillInlineMenuIframeService.initMenuIframe();

      expect(autofillInlineMenuIframeService["iframe"]).toMatchSnapshot();
    });

    it("appends the iframe to the shadowDom", () => {
      jest.spyOn(autofillInlineMenuIframeService["shadow"], "appendChild");

      autofillInlineMenuIframeService.initMenuIframe();

      expect(autofillInlineMenuIframeService["shadow"].appendChild).toHaveBeenCalledWith(
        autofillInlineMenuIframeService["iframe"],
      );
    });

    // TODO CG - This test is brittle and failing due to how we are calling the private method. This needs to be reworked
    it.skip("creates an aria alert element if the ariaAlert param is passed", () => {
      const ariaAlert = "aria alert";
      jest.spyOn(autofillInlineMenuIframeService as any, "createAriaAlertElement");

      autofillInlineMenuIframeService.initMenuIframe();

      expect(autofillInlineMenuIframeService["createAriaAlertElement"]).toHaveBeenCalledWith(
        ariaAlert,
      );
      expect(autofillInlineMenuIframeService["ariaAlertElement"]).toMatchSnapshot();
    });

    describe("on load of the iframe source", () => {
      beforeEach(() => {
        autofillInlineMenuIframeService.initMenuIframe();
      });

      it("sets up and connects the port message listener to the extension background", () => {
        jest.spyOn(globalThis, "addEventListener");

        autofillInlineMenuIframeService["iframe"].dispatchEvent(new Event(EVENTS.LOAD));
        portSpy = autofillInlineMenuIframeService["port"];

        expect(chrome.runtime.connect).toHaveBeenCalledWith({ name: AutofillOverlayPort.Button });
        expect(portSpy.onDisconnect.addListener).toHaveBeenCalledWith(handlePortDisconnectSpy);
        expect(portSpy.onMessage.addListener).toHaveBeenCalledWith(handlePortMessageSpy);
      });

      it("skips announcing the aria alert if the aria alert element is not populated", () => {
        jest.spyOn(globalThis, "setTimeout");
        autofillInlineMenuIframeService["ariaAlertElement"] = undefined;

        autofillInlineMenuIframeService["iframe"].dispatchEvent(new Event(EVENTS.LOAD));

        expect(globalThis.setTimeout).not.toHaveBeenCalled();
      });

      it("announces the aria alert if the aria alert element is populated", async () => {
        jest.useFakeTimers();
        jest.spyOn(globalThis, "setTimeout");
        sendExtensionMessageSpy.mockResolvedValue(true);
        autofillInlineMenuIframeService["ariaAlertElement"] = document.createElement("div");
        autofillInlineMenuIframeService["ariaAlertTimeout"] = setTimeout(jest.fn(), 2000);

        autofillInlineMenuIframeService["iframe"].dispatchEvent(new Event(EVENTS.LOAD));

        expect(globalThis.setTimeout).toHaveBeenCalled();
        jest.advanceTimersByTime(2000);
        await flushPromises();

        expect(shadowAppendSpy).toHaveBeenCalledWith(
          autofillInlineMenuIframeService["ariaAlertElement"],
        );
      });
    });
  });

  describe("event listeners", () => {
    beforeEach(() => {
      autofillInlineMenuIframeService.initMenuIframe();
      autofillInlineMenuIframeService["iframe"].dispatchEvent(new Event(EVENTS.LOAD));
      Object.defineProperty(autofillInlineMenuIframeService["iframe"], "contentWindow", {
        value: {
          postMessage: jest.fn(),
        },
        writable: true,
      });
      jest.spyOn(autofillInlineMenuIframeService["iframe"].contentWindow, "postMessage");
      portSpy = autofillInlineMenuIframeService["port"];
    });

    describe("handlePortDisconnect", () => {
      it("ignores ports that do not have the correct port name", () => {
        portSpy.name = "wrong-port-name";
        triggerPortOnDisconnectEvent(portSpy);

        expect(autofillInlineMenuIframeService["port"]).not.toBeNull();
      });

      it("resets the iframe element's opacity, height, and display styles", () => {
        triggerPortOnDisconnectEvent(portSpy);

        expect(autofillInlineMenuIframeService["iframe"].style.opacity).toBe("0");
        expect(autofillInlineMenuIframeService["iframe"].style.height).toBe("0px");
        expect(autofillInlineMenuIframeService["iframe"].style.display).toBe("block");
      });

      it("removes the port's onMessage listener", () => {
        triggerPortOnDisconnectEvent(portSpy);

        expect(portSpy.onMessage.removeListener).toHaveBeenCalledWith(handlePortMessageSpy);
      });

      it("removes the port's onDisconnect listener", () => {
        triggerPortOnDisconnectEvent(portSpy);

        expect(portSpy.onDisconnect.removeListener).toHaveBeenCalledWith(handlePortDisconnectSpy);
      });

      it("disconnects the port", () => {
        triggerPortOnDisconnectEvent(portSpy);

        expect(portSpy.disconnect).toHaveBeenCalled();
        expect(autofillInlineMenuIframeService["port"]).toBeNull();
      });
    });

    describe("handlePortMessage", () => {
      it("ignores port messages that do not correlate to the correct port name", () => {
        portSpy.name = "wrong-port-name";
        sendPortMessage(portSpy, {});

        expect(
          autofillInlineMenuIframeService["iframe"].contentWindow.postMessage,
        ).not.toHaveBeenCalled();
      });

      it("passes on the message to the iframe if the message is not registered with the message handlers", () => {
        const message = { command: "unregisteredMessage" };

        sendPortMessage(portSpy, message);

        expect(
          autofillInlineMenuIframeService["iframe"].contentWindow.postMessage,
        ).toHaveBeenCalledWith(message, "*");
      });

      it("handles port messages that are registered with the message handlers and does not pass the message on to the iframe", () => {
        jest.spyOn(autofillInlineMenuIframeService as any, "updateIframePosition");

        sendPortMessage(portSpy, { command: "updateAutofillInlineMenuPosition" });

        expect(
          autofillInlineMenuIframeService["iframe"].contentWindow.postMessage,
        ).not.toHaveBeenCalled();
      });

      describe("initializing the inline menu button", () => {
        it("sets the port key and posts the message to the inline menu page iframe", () => {
          const portKey = "portKey";
          const message = {
            command: "initAutofillInlineMenuButton",
            portKey,
          };

          sendPortMessage(portSpy, message);

          expect(autofillInlineMenuIframeService["portKey"]).toBe(portKey);
          expect(
            autofillInlineMenuIframeService["iframe"].contentWindow.postMessage,
          ).toHaveBeenCalledWith(message, "*");
        });
      });

      describe("initializing the inline menu list", () => {
        let updateElementStylesSpy: jest.SpyInstance;

        beforeEach(() => {
          updateElementStylesSpy = jest.spyOn(
            autofillInlineMenuIframeService as any,
            "updateElementStyles",
          );
        });

        it("passes the message on to the iframe element", () => {
          const message = {
            command: "initAutofillInlineMenuList",
            theme: ThemeType.Light,
          };

          sendPortMessage(portSpy, message);

          expect(updateElementStylesSpy).not.toHaveBeenCalled();
          expect(
            autofillInlineMenuIframeService["iframe"].contentWindow.postMessage,
          ).toHaveBeenCalledWith(message, "*");
        });

        it("sets a light theme based on the user's system preferences", () => {
          window.matchMedia = jest.fn(() => mock<MediaQueryList>({ matches: false }));
          const message = {
            command: "initAutofillInlineMenuList",
            theme: ThemeType.System,
          };

          sendPortMessage(portSpy, message);

          expect(window.matchMedia).toHaveBeenCalledWith("(prefers-color-scheme: dark)");
          expect(
            autofillInlineMenuIframeService["iframe"].contentWindow.postMessage,
          ).toHaveBeenCalledWith(
            {
              command: "initAutofillInlineMenuList",
              theme: ThemeType.Light,
            },
            "*",
          );
        });

        it("sets a dark theme based on the user's system preferences", () => {
          window.matchMedia = jest.fn(() => mock<MediaQueryList>({ matches: true }));
          const message = {
            command: "initAutofillInlineMenuList",
            theme: ThemeType.System,
          };

          sendPortMessage(portSpy, message);

          expect(window.matchMedia).toHaveBeenCalledWith("(prefers-color-scheme: dark)");
          expect(
            autofillInlineMenuIframeService["iframe"].contentWindow.postMessage,
          ).toHaveBeenCalledWith(
            {
              command: "initAutofillInlineMenuList",
              theme: ThemeType.Dark,
            },
            "*",
          );
        });

        it("updates the border to match the `dark` theme", () => {
          const message = {
            command: "initAutofillInlineMenuList",
            theme: ThemeType.Dark,
          };

          sendPortMessage(portSpy, message);

          expect(updateElementStylesSpy).toHaveBeenCalledWith(
            autofillInlineMenuIframeService["iframe"],
            {
              borderColor: "#4c525f",
            },
          );
        });
      });

      describe("updating the iframe's position", () => {
        beforeEach(() => {
          jest.spyOn(globalThis.document, "hasFocus").mockReturnValue(true);
        });

        it("ignores updating the iframe position if the document does not have focus", () => {
          jest.spyOn(autofillInlineMenuIframeService as any, "updateElementStyles");
          jest.spyOn(globalThis.document, "hasFocus").mockReturnValue(false);

          sendPortMessage(portSpy, {
            command: "updateAutofillInlineMenuPosition",
            styles: { top: 100, left: 100 },
          });

          expect(autofillInlineMenuIframeService["updateElementStyles"]).not.toHaveBeenCalled();
        });

        it("updates the iframe position if the document has focus", () => {
          const styles = { top: "100px", left: "100px" };

          sendPortMessage(portSpy, {
            command: "updateAutofillInlineMenuPosition",
            styles,
          });

          expect(autofillInlineMenuIframeService["iframe"].style.top).toBe(styles.top);
          expect(autofillInlineMenuIframeService["iframe"].style.left).toBe(styles.left);
        });

        it("announces the opening of the iframe using an aria alert", async () => {
          jest.useFakeTimers();
          sendExtensionMessageSpy.mockResolvedValue(true);
          const styles = { top: "100px", left: "100px" };

          sendPortMessage(portSpy, {
            command: "updateAutofillInlineMenuPosition",
            styles,
          });
          jest.advanceTimersByTime(2000);
          await flushPromises();

          expect(shadowAppendSpy).toHaveBeenCalledWith(
            autofillInlineMenuIframeService["ariaAlertElement"],
          );
        });

        it("resets the fade in timeout if it is set", () => {
          autofillInlineMenuIframeService["fadeInTimeout"] = setTimeout(jest.fn, 100);
          const styles = { top: "100px", left: "100px" };
          jest.spyOn(autofillInlineMenuIframeService as any, "handleFadeInInlineMenuIframe");

          sendPortMessage(portSpy, {
            command: "updateAutofillInlineMenuPosition",
            styles,
          });

          expect(
            autofillInlineMenuIframeService["handleFadeInInlineMenuIframe"],
          ).toHaveBeenCalled();
        });
      });

      it("updates the visibility of the iframe", () => {
        sendPortMessage(portSpy, {
          command: "toggleAutofillInlineMenuHidden",
          styles: { display: "none" },
        });

        expect(autofillInlineMenuIframeService["iframe"].style.display).toBe("none");
      });

      it("updates the button based on the web page's color scheme", () => {
        sendPortMessage(portSpy, {
          command: "updateAutofillInlineMenuColorScheme",
        });

        expect(
          autofillInlineMenuIframeService["iframe"].contentWindow.postMessage,
        ).toHaveBeenCalledWith(
          {
            command: "updateAutofillInlineMenuColorScheme",
            colorScheme: "normal",
          },
          "*",
        );
      });

      it("triggers a delayed closure of the inline menu", () => {
        jest.useFakeTimers();
        jest.spyOn(globalThis, "clearTimeout");
        autofillInlineMenuIframeService["delayedCloseTimeout"] = setTimeout(jest.fn, 100);

        sendPortMessage(portSpy, { command: "triggerDelayedAutofillInlineMenuClosure" });
        expect(clearTimeout).toHaveBeenCalled();
        expect(autofillInlineMenuIframeService["iframe"].style.opacity).toBe("0");
        expect(autofillInlineMenuIframeService["iframe"].style.transition).toBe(
          "opacity 65ms ease-out 0s",
        );

        jest.advanceTimersByTime(100);
        expect(autofillInlineMenuIframeService["iframe"].style.transition).toBe(
          "opacity 125ms ease-out 0s",
        );
        expect(sendExtensionMessageSpy).toHaveBeenCalledWith("closeAutofillInlineMenu", {
          forceCloseInlineMenu: true,
        });
      });

      it("triggers a fade in of the inline menu", () => {
        jest.useFakeTimers();
        jest.spyOn(globalThis, "clearTimeout");
        autofillInlineMenuIframeService["fadeInTimeout"] = setTimeout(jest.fn, 10);

        sendPortMessage(portSpy, { command: "fadeInAutofillInlineMenuIframe" });
        expect(clearTimeout).toHaveBeenCalled();
        expect(autofillInlineMenuIframeService["iframe"].style.opacity).toBe("0");

        jest.advanceTimersByTime(10);
        expect(autofillInlineMenuIframeService["iframe"].style.opacity).toBe("1");
      });

      it("triggers an aria alert when the password in regenerated", () => {
        jest.spyOn(autofillInlineMenuIframeService as any, "createAriaAlertElement");

        sendPortMessage(portSpy, {
          command: "updateAutofillInlineMenuGeneratedPassword",
          refreshPassword: true,
        });

        expect(autofillInlineMenuIframeService["createAriaAlertElement"]).toHaveBeenCalledWith(
          true,
        );
      });
    });
  });

  describe("mutation observer", () => {
    beforeEach(() => {
      autofillInlineMenuIframeService.initMenuIframe();
      autofillInlineMenuIframeService["iframe"].dispatchEvent(new Event(EVENTS.LOAD));
      portSpy = autofillInlineMenuIframeService["port"];
    });

    it("skips handling found mutations if excessive mutations are triggering", async () => {
      jest.useFakeTimers();
      jest
        .spyOn(
          autofillInlineMenuIframeService as any,
          "isTriggeringExcessiveMutationObserverIterations",
        )
        .mockReturnValue(true);
      jest.spyOn(autofillInlineMenuIframeService as any, "updateElementStyles");

      autofillInlineMenuIframeService["iframe"].style.visibility = "hidden";
      await flushPromises();

      expect(autofillInlineMenuIframeService["updateElementStyles"]).not.toHaveBeenCalled();
    });

    it("reverts any styles changes made directly to the iframe", async () => {
      jest.useFakeTimers();

      autofillInlineMenuIframeService["iframe"].style.visibility = "hidden";
      await flushPromises();

      expect(autofillInlineMenuIframeService["iframe"].style.visibility).toBe("visible");
    });

    it("force closes the autofill inline menu if more than 9 foreign mutations are triggered", async () => {
      jest.useFakeTimers();
      autofillInlineMenuIframeService["foreignMutationsCount"] = 10;

      autofillInlineMenuIframeService["iframe"].src = "http://malicious-site.com";
      await flushPromises();

      expect(sendExtensionMessageSpy).toHaveBeenCalledWith("closeAutofillInlineMenu", {
        forceCloseInlineMenu: true,
      });
    });

    it("force closes the autofill overinline menulay if excessive mutations are being triggered", async () => {
      jest.useFakeTimers();
      autofillInlineMenuIframeService["mutationObserverIterations"] = 20;

      autofillInlineMenuIframeService["iframe"].src = "http://malicious-site.com";
      await flushPromises();

      expect(sendExtensionMessageSpy).toHaveBeenCalledWith("closeAutofillInlineMenu", {
        forceCloseInlineMenu: true,
      });
    });

    it("resets the excessive mutations and foreign mutation counters", async () => {
      jest.useFakeTimers();
      autofillInlineMenuIframeService["foreignMutationsCount"] = 9;
      autofillInlineMenuIframeService["mutationObserverIterations"] = 19;

      autofillInlineMenuIframeService["iframe"].src = "http://malicious-site.com";
      jest.advanceTimersByTime(2001);
      await flushPromises();

      expect(autofillInlineMenuIframeService["foreignMutationsCount"]).toBe(0);
      expect(autofillInlineMenuIframeService["mutationObserverIterations"]).toBe(0);
    });

    it("resets any mutated default attributes for the iframe", async () => {
      jest.useFakeTimers();

      autofillInlineMenuIframeService["iframe"].title = "some-other-title";
      await flushPromises();

      expect(autofillInlineMenuIframeService["iframe"].title).toBe("title");
    });
  });
});
