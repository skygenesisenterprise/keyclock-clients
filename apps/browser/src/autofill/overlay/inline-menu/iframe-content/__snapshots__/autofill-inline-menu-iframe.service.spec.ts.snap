// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutofillInlineMenuIframeService initMenuIframe sets up the iframe's attributes 1`] = `
<iframe
  allowtransparency="true"
  scrolling="no"
  src="chrome-extension://id/overlay/menu.html"
  style="all: initial !important; position: fixed !important; display: block !important; z-index: 2147483647 !important; line-height: 0 !important; overflow: hidden !important; transition: opacity 125ms ease-out 0s !important; visibility: visible !important; clip-path: none !important; pointer-events: auto !important; margin: 0px !important; padding: 0px !important; color-scheme: normal !important; opacity: 0 !important; height: 0px;"
  tabindex="-1"
  title="title"
/>
`;
